### PR TITLE
Return redirect error to redirect, and can customize handle normal error

### PR DESCRIPTION
**Pull request by sunfmin**

### CODE REVIEW

**Comment by sunfmin** [comment]
@JadenTeng 搞定。

**Comment by bodhi** [comment]
@sunfmin you want to allow specific containers to redirect the whole request? Why isn't the top-level HTTP handler taking care of this? Containers should just be for *rendering data*, no?

**Comment by sunfmin** [comment]
A lots of specific logic depends on the data that get inside from the container. If we do it in middleware, We have to get those data in the middleware from the beginning. and duplicate that logic inside container.. Or pass it in. If we passing data into containers, Then container is no difference with gorazor templates...

**Comment by bodhi** [comment]
Ok, but why can't an application specific page handler handle this case?

The problem I see is that there's no way to know from the signature of a container whether the container could ask to redirect instead of rendering. This means you need to wrap *every* handler with a redirection error handler.

It also will prevent streaming containers, you are forced to render everything and keep it as a string, before you can send it back to the browser. This isn't a major problem, but something to consider.

Also, can you explain why the middleware is needed?

**Comment by sunfmin** [comment]
I mean middleware to the same as Handler... We now don't have handler for each URL, they are sharing the same one now.

File: errors.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package containers
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+)
+
+type ErrHandler interface {
```
Comment (sunfmin) [inline errors.go]: > This means you need to wrap every handler with a redirection error handler.

This pull request added a content of ErrHandler interface, any error that implements the ErrHandler interface, will handle it self.

File: errors.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	h  http.Handler
+	eh ErrHandler
+}
+
+func (iseh *internalServerErrorHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	r = r.WithContext(context.WithValue(r.Context(), internalServerErrorHandlerKey, iseh.eh))
+	iseh.h.ServeHTTP(w, r)
+}
+
+func UseErrHandler(h http.Handler, errhandler ErrHandler) (handler http.Handler) {
```
Comment (sunfmin) [inline errors.go]: And We have this middleware you can wrap a default ErrHandler to your Handler.

Comment (sunfmin) [inline errors.go]: You don't need to wrap this, If you just return an error with `NewRedirectError`

File: page_handler.go [hunk 3]
```diff
@@ -16,6 +16,17 @@ func PageHandler(page Page, layout Layout) http.Handler {
...
+		return
+	}
+
+	eh, ok := r.Context().Value(internalServerErrorHandlerKey).(ErrHandler)
+	if ok {
+		eh.HandleErr(w, r, err)
+		return
+	}
+
 	w.WriteHeader(http.StatusInternalServerError)
```
Comment (sunfmin) [inline page_handler.go]: The previous implementation of `handleError` is kind of pale, And not extendable.

**Comment by bodhi** [comment]
Hold your horses! Why'd you merge this?

**Comment by bodhi** [comment]
Is the context necessary? Can you just pass an error handler in when you set up the initial page handlers?

**Comment by sunfmin** [comment]
We can have a conversation after aigle launch. But right now we need this for aigle to launch.

**Comment by bodhi** [comment]
> But right now we need this for aigle to launch.

Next time please lock to your PR branch instead of master in this case.

**Pull request merged**